packages/vite/src/node/ssr/ssrExternal.ts

```diff
@@ -86,8 +88,97 @@ const CJS_CONTENT_RE =
 // TODO: use import()
 const _require = createRequire(import.meta.url)
 
-// do we need to do this ahead of time or could we do it lazily?
-function collectExternals(
+const isSsrExternalCache = new WeakMap<
+  ResolvedConfig,
+  (id: string) => boolean | undefined
+>()
+
+export function shouldExternalizeForSSR(
+  id: string,
+  config: ResolvedConfig
+): boolean | undefined {
+  let isSsrExternal = isSsrExternalCache.get(config)
+  if (!isSsrExternal) {
+    isSsrExternal = createIsSsrExternal(config)
+    isSsrExternalCache.set(config, isSsrExternal)
+  }
+  return isSsrExternal(id)
+}
+
+function createIsSsrExternal(
+  config: ResolvedConfig
+): (id: string) => boolean | undefined {
+  const processedIds = new Map<string, boolean | undefined>()
+
+  const { ssr, root } = config
+
+  const noExternal = ssr?.noExternal
+  const noExternalFilter =
+    noExternal !== 'undefined' &&
+    typeof noExternal !== 'boolean' &&
+    createFilter(undefined, noExternal, { resolve: false })
+
+  const isConfiguredAsExternal = (id: string) => {
+    const { ssr } = config
+    if (!ssr || ssr.external?.includes(id)) {
+      return true
+    }
+    if (typeof noExternal === 'boolean') {
+      return !noExternal
+    }
+    if (noExternalFilter) {
+      return noExternalFilter(id)
+    }
+    return true
+  }
+
+  const resolveOptions: InternalResolveOptions = {
+    root,
+    preserveSymlinks: config.resolve.preserveSymlinks,
+    isProduction: false,
+    isBuild: true
+  }
+
+  const isPackageEntry = (id: string) => {
+    if (!bareImportRE.test(id) || id.includes('\0')) {
+      return false
+    }
+    if (
+      tryNodeResolve(
+        id,
+        undefined,
+        resolveOptions,
+        ssr?.target === 'webworker',
+        undefined,
+        true
+      )
+    ) {
+      return true
+    }
+    try {
+      // no main entry, but deep imports may be allowed
+      const pkgPath = resolveFrom(`${id}/package.json`, root)
+      if (pkgPath.includes('node_modules')) {
```

Comment:
Do we really want to discriminate between installed packages and linked packages (e.g. `npm link some-package)`? Isn't the presence of `package.json` a strong enough signal that this is pacakge?

Reply:
Yes, I copied this code from the previous logic, but now it is a good time to change things if we see fit. We can try to simplify this even more and then add back logic if there are issues.

Reply:
See https://github.com/vitejs/vite/pull/8348#issuecomment-1140305825